src/webserver.js

```diff
@@ -0,0 +1,263 @@
+ */
+class WebServer {
+    /**
+     * @param {string} hostname The hostname the server listens on
+     * @param {number} port The port the server listens on
+     * @param {string} wwwPath The path to serve static files from
+     * @param {string} indexFile The file used when the client requests "/"
+     */
+    constructor(hostname, port, wwwPath, indexFile = "/index.html") {
+        typeassert.assertStringNotEmpty(hostname);
+        typeassert.assertInteger(port);
+        typeassert.assertStringNotEmpty(wwwPath);
+        typeassert.assertStringNotEmpty(indexFile);
+
+        this.hostname = hostname;
+        this.port = port;
+
+
+
+        this.server = http.createServer(this.requestListener);
+        this.server.wwwPath = wwwPath;
+        this.server.getHandlers = new Map();
+        this.server.postHandlers = new Map();
+        this.server.indexFile = indexFile;
+
+        Object.freeze(this);
+    }
+
+    /**
+     * @summary Add a get handler
+     * @param {string} url The url to handle
+     * @param {getCallback} handler The handler
+     */
+    addGetHandler(url, handler) {
+        typeassert.assertStringNotEmpty(url);
+        typeassert.assertFunction(handler);
+
+        this.server.getHandlers.set(url, handler);
+    }
+
+    /**
+     * @summary Add a post handler
+     * @param {string} url The url to handle
+     * @param {postCallback} handler The handler
+     */
+    addPostHandler(url, handler) {
+        typeassert.assertStringNotEmpty(url);
+        typeassert.assertFunction(handler);
+
+        this.server.postHandlers.set(url, handler);
+    }
+
+
+    /**
+     * @summary Run the server
+     */
+    run() {
+        return new Promise((resolve) => {
+            this.server.listen(this.port, this.hostname, () => resolve());
+        });
+    }
+
+    stop() {
+        return new Promise((resolve) => {
+            this.server.close(() => resolve());
+        });
+    }
+
+
+    /**
+     * @param {http.IncomingMessage} request The http request
+     * @param {http.ServerResponse} response The servers respond to the request
+     */
+    requestListener(request, response) {
+        const thiz = this;
+        /**
+         * @param {number} code The http status code
+         */
+        function errorResponse(code) {
+            typeassert.assertInteger(code);
+
+            response.statusCode = code;
+
+            // Mime type
+            response.setHeader("Content-Type", mimeType.contentType("text/txt"));
+
+            response.write(http.STATUS_CODES[`${code}`]);
+            response.end("\n");
+        }
+
+        /**
+         * @returns {Promise} Post body data in form of URLSearchParams
+         */
+        function extractForm() {
+            if (request.headers["content-type"] === undefined) {
+                return Promise.resolve(null);
+            }
+
+            const ct = contentType.parse(request);
+
+            if (ct.type !== "application/x-www-form-urlencoded") {
```

Comment:
Browseren kan godt finde på at sende form data på andre måder. Sørg for at det virker